### ISSUE TITLE
Which words are reserved under ViewComponent?

### ISSUE THREAD

**Issue by boardfish** [github/view_component#976]
I've just encountered an issue whereby having a slot like `renders_many :tags` results in a conflict between the `tag` slot and the `tag` view helper. It'd be good to keep track of which other words, if any, are likely to cause issues with ViewComponent.

**Comment by joelhawksley** [comment]
@boardfish I agree. Our work on reducing the number of potential instance variable clashes was a good step in this direction. How do you think we should keep track of these words?

**Comment by boardfish** [comment]
I wonder if tests would be a good way to do this. In the same way that the test for the trailing newline issue uses something peripheral to ViewComponent to document the solution to the issue, we could do something similar for namespace clashes. 

For example, in the case I had above, we could have passing tests to suggest alternative names like `topics` or `categories` for the slot. `attributes` is another namespace clash, now that I think about it, if you're using Active Model on top of your base component class.

**Comment by joelhawksley** [comment]
@boardfish hmm. I wonder if we could enforce this at the framework level. For example, we could check slot names against a list of reserved words at compile time and raise an error.

**Comment by boardfish** [comment]
That could be really good for addressing this in a way that's more developer-friendly.

I wonder if reserved words could affect the experience in other ways, perhaps such as:
- defining a method that's the same as a slot name
- doing something similar with some of ViewComponent's convenience features like collection parameters
- accidentally overriding private methods
I'm interested in experimenting with these to see what kinds of situations can arise.

I suppose the logic we'd use to raise errors in case of reserved words would be common to all these cases, but the words themselves may differ between each case.

**Comment by joelhawksley** [comment]
`accidentally overriding private methods`

Oh, that is an interesting one. I hadn't thought of namespacing our private methods.

Would you be up for taking on the slot naming issue and seeing what you can come up with?

**Comment by boardfish** [comment]
Yeah, I'll see what I can whip up there.

**Comment by boardfish** [comment]
Turns out there's a validation in place for if `renders_one :content` is called, but there isn't for `renders_many :contents`, so I'll try to address that in isolation first.

I'll see how it interfaces with other things too, but this is one well-defined area we can tackle for sure.

**Comment by boardfish** [comment]
I've had a look at this just now - I've tried using slots to override existing methods at various levels. Not sure how confident I feel in this, so if anyone's got more knowledge of how to break Ruby and libraries, I'd be interested to know if there's anything else I can try around here. So far I've tried slots named `text_field_tag` (a view helper), `freeze` (a method on the component), `methods`, and `initialize_parameters` (a private method on the component). I've also tried calling these via a template and via `call`. These don't seem to mess with rendering the component at all.

I guess the usual warning of "avoid overriding library methods" flies, but I don't think there's really much more we can do here without evidence of a conflict. `content` was something there was evidence for, but I can't really think of anything else that would trigger this. I think I'll close this, but I'm willing to investigate if something else to this effect comes up.
